### PR TITLE
Pr use MultiDiscrete instead of DiscreteSequence so gym _spec_from_gym_space won't complain

### PR DESCRIPTION
**Pull request by le-horizon**
Make groceryground more configurable.  Update input type from DiscreteSequence to MultiDiscrete for language task (because tf_agents only supports using their predefined types during _spec_from_gym_space).

### CODE REVIEW

**Comment by emailweixu** [comment]
Conflicts need to be resolved. The title of this PR does not seem right.

**Comment by le-horizon** [comment]
Oops, forgot to pull before requesting review..  Resolved conflict, and tested image+language training.

Could you take another look?

Thanks,
Le

File: python/social_bot/__init__.py [hunk 1]
```diff
@@ -57,6 +63,12 @@
     max_episode_steps=200,
 )
 
+register(
+    id='SocialBot-GroceryGroundImageLanguage-v1',
+    entry_point='social_bot.envs:GroceryGroundImageLanguage',
+    max_episode_steps=1e8,
```
Comment (emailweixu) [inline python/social_bot/__init__.py]: There is no need to register another one if the difference with v0 is only max_episode_steps, max_episode_steps can be configured by suite_gym.load.max_episode_steps=1000

File: python/social_bot/envs/grocery_ground.py [hunk 2]
```diff
@@ -434,6 +435,9 @@ def __init__(self,
                  image_with_internal_states=False,
                  task_name='goal',
                  agent_type='pioneer2dx_noplugin',
+                 random_goal=None,
+                 fail_distance_thresh=3,
+                 max_steps=200,
```
Comment (emailweixu) [inline python/social_bot/envs/grocery_ground.py]: all of these can be configured through GroceryGroundGoalTask

File: python/social_bot/envs/grocery_ground.py [hunk 3]
```diff
@@ -611,6 +625,8 @@ def step(self, action):
             if type(sentence) != str:
                 sentence = self._teacher.sequence_to_sentence(sentence)
             action_ctrl = action['control']
+            #if not self._train:
+            #    tf.compat.v1.logging.info("Teacher sentence: " + sentence)
```
Comment (emailweixu) [inline python/social_bot/envs/grocery_ground.py]: remove

File: python/social_bot/__init__.py (hunk 1 above)
Comment (le-horizon) [reply python/social_bot/__init__.py]: Good to know!

File: python/social_bot/envs/grocery_ground.py (hunk 2 above)
Comment (le-horizon) [reply python/social_bot/envs/grocery_ground.py]: Oh, somehow missed that.

Comment (le-horizon) [reply python/social_bot/envs/grocery_ground.py]: Actually these are fixed around line 497 when initializing the GroceryGroundGoalTask.  This change makes it configurable.

**Comment by le-horizon** [comment]
updated code again.  Some of the changes to make gin configurable is still needed..

File: python/social_bot/envs/grocery_ground.py [hunk 4]
```diff
@@ -86,6 +86,8 @@ def __init__(self,
             random_range (float): the goal's random position range
             random_goal (bool): if ture, teacher will randomly select goal from the object list each episode
         """
+        if goal_name is None:
```
Comment (Jialn) [inline python/social_bot/envs/grocery_ground.py]: it seems this can be removed

File: python/social_bot/envs/grocery_ground.py (hunk 2 above)
Comment (Jialn) [reply python/social_bot/envs/grocery_ground.py]: I also think it is better to configure through gin of GroceryGroundGoalTask.  GroceryGround is meant for many task, some of them does not need random_goal or fail_distance_thresh.  We can removed the parameters when initializing the task at line 497.

Comment (Jialn) [reply python/social_bot/envs/grocery_ground.py]: As to max_steps,  it should also be removed here. We have the gym parameter max_episode_steps so don't bother with this parameter. And most time we should use max_episode_steps, so that the final return of the last step can be correctly handled:
If max_episode_steps is reached and the env is reset, the potential discounted future reward will be take into account when update the value network. Most task should be like this. 
If max_steps is reached with 'done' and the env is reset, the algorithm don't know env reaches time limits and treat this the same as done. All potential discounted reward in the future will be considered as zero. For some task like goal task, that also could work. but using max_episode_steps is better.
If max_steps is reached without 'done' and the env is reset, the algorithm part don't know env is reset, the initial state of next episode will be considered as the next state of the last step. This should never happen.

You can configure the max_episode_steps like this: suite_socialbot.load.max_episode_steps=300

File: python/social_bot/envs/grocery_ground.py [hunk 5]
```diff
@@ -480,11 +492,13 @@ def __init__(self,
 
         self._teacher = teacher.Teacher(task_groups_exclusive=False)
         if task_name is None or task_name == 'goal':
+            if random_goal is None:
```
Comment (Jialn) [inline python/social_bot/envs/grocery_ground.py]: this can be removed

File: python/social_bot/envs/grocery_ground.py (hunk 4 above)
Comment (le-horizon) [reply python/social_bot/envs/grocery_ground.py]: Good observation Jiangtao.  I put it here just in case the user passes in `goal_name = None` in the gin file, in which case, a default value is still provided here.

Comment (le-horizon) [reply python/social_bot/envs/grocery_ground.py]: Maybe it's better to alert the user if that happens?

File: python/social_bot/envs/grocery_ground.py [hunk 6]
```diff
@@ -480,11 +492,13 @@ def __init__(self,
...
+                random_goal = with_language
             main_task = GroceryGroundGoalTask(
-                max_steps=200,
+                max_steps=max_steps,
```
Comment (Jialn) [inline python/social_bot/envs/grocery_ground.py]: We can remove 498-502, and make the default parameter of GroceryGroundGoalTask the same as the initialization here. 
We can config GroceryGroundGoalTask through gin when needed.

File: python/social_bot/envs/grocery_ground.py (hunk 4 above)
Comment (Jialn) [reply python/social_bot/envs/grocery_ground.py]: It's ok just remove it. No one would set goal_name=None in gin file...

File: python/social_bot/envs/grocery_ground.py [hunk 7]
```diff
@@ -70,7 +70,7 @@ class GroceryGroundGoalTask(GroceryGroundTaskBase, GoalTask):
 
     def __init__(self,
```
Comment (Jialn) [inline python/social_bot/envs/grocery_ground.py]: The default "__init__()" parameter could be modified to be the same as initializing code, to keep the default behavior the same after removed the initialing code.

File: python/social_bot/envs/grocery_ground.py (hunk 5 above)
Comment (le-horizon) [reply python/social_bot/envs/grocery_ground.py]: Done.

File: python/social_bot/envs/grocery_ground.py (hunk 6 above)
Comment (le-horizon) [reply python/social_bot/envs/grocery_ground.py]: Good point!  Done.

File: python/social_bot/envs/grocery_ground.py (hunk 7 above)
Comment (le-horizon) [reply python/social_bot/envs/grocery_ground.py]: Done.

**Comment by le-horizon** [comment]
Good points Jiangtao and good to know about the complexity involved in stopping episodes.
Done as you suggested.  Could you take another look?

Thanks,
Le

File: python/social_bot/envs/grocery_ground.py [hunk 8]
```diff
@@ -500,9 +497,13 @@ def __init__(self,
             icub_standing_task = ICubStandingTask()
             icub_aux_task_group.add_task(icub_standing_task)
             self._teacher.add_task_group(icub_aux_task_group)
+        self._teacher._build_vocab_from_tasks()
```
Comment (Jialn) [inline python/social_bot/envs/grocery_ground.py]: This might need to be removed. _build_vocab_from_tasks is done within teacher.py automatically when "teacher.sentence_to_sequence()" is called

Comment (le-horizon) [reply python/social_bot/envs/grocery_ground.py]: I need the vocab_size before sentence_to_sequence() is called.  Unless I missed something, this call is needed?

Comment (Jialn) [reply python/social_bot/envs/grocery_ground.py]: Sorry. I have not noticed that : D

**Comment by Jialn** [comment]
Is there any chance to import DiscreteSequence to alf, and make it support language task? Seems multi-discrete action is made for game controllers or keyboards. I think it would be better to have a gym-space designed for language, and is convenient to extend to other encoding/embedding in the future, like sub-words or something else.

**Comment by emailweixu** [comment]
MultiDiscrete seems ok. If we change to use pretrained continuous word embedding, we can use Box space.

> Is there any chance to import DiscreteSequence to alf, and make it support language task? Seems multi-discrete action is made for game controllers or keyboards. I think it would be better to have a gym-space designed for language, and is convenient to extend to other encoding/embedding in the future, like sub-words or something else.

File: python/social_bot/envs/grocery_ground.py (hunk 2 above)
Comment (le-horizon) [reply python/social_bot/envs/grocery_ground.py]: Good suggestions.  All done.

**Comment by le-horizon** [comment]
> Is there any chance to import DiscreteSequence to alf, and make it support language task? Seems multi-discrete action is made for game controllers or keyboards. I think it would be better to have a gym-space designed for language, and is convenient to extend to other encoding/embedding in the future, like sub-words or something else.

Unfortunately it's not an alf requirement.  It's tf_agents requiring only supported classes: https://github.com/tensorflow/agents/blob/master/tf_agents/environments/gym_wrapper.py#L107

**Comment by Jialn** [comment]
OK. Seems we have no choice   :-)
Shall we remove DiscreteSequence and use multi-discrete if it is not useful?

**Comment by le-horizon** [comment]
Hi Wei, still need your approval.

For DiscreteSequence, I think we at least need to make a comment saying currently tf_agents requires us to use MultiDiscrete.  Maybe it will be supported in the future?
